### PR TITLE
Implement @main functions

### PR DESCRIPTION
**Pull request by odersky**
See: 

https://github.com/dotty-staging/dotty/blob/add-main-fn/docs/docs/reference/changed-features/main-functions.md

### CODE REVIEW

File: docs/docs/reference/changed-features/main-functions.md [hunk 1]
```diff
@@ -0,0 +1,82 @@
+---
+layout: doc-page
+title: "Main Methods"
+---
+
+Scala 3 offers a new way to define programs that can be invoked from the command line:
+A @main annotation on a method turns this method into an executable program.
```
Comment (nafg) [inline docs/docs/reference/changed-features/main-functions.md]: missing backticks

Comment (nafg) [inline docs/docs/reference/changed-features/main-functions.md]: missing backticks

File: library/src/scala/main.scala [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+**  __\ \/ /__/ __ |/ /__/ __ |    http://scala-lang.org/               **
+** /____/\___/_/ |_/____/_/ | |                                         **
+**                          |/                                          **
+\*                                                                      */
+
+package scala
+
+/** An annotation that designates a main function
+ */
+class main extends scala.annotation.Annotation {}
```
Comment (hepin1989) [inline library/src/scala/main.scala]: final

Comment (odersky) [reply library/src/scala/main.scala]: 👍

File: docs/docs/reference/changed-features/main-functions.md (hunk 1 above)
Comment (odersky) [reply docs/docs/reference/changed-features/main-functions.md]: thanks!

File: compiler/src/dotty/tools/dotc/ast/MainProxies.scala [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+
+import core._
+import Symbols._, Types._, Contexts._, Decorators._, util.Spans._, Flags._, Constants._
+import StdNames.nme
+import ast.Trees._
+
+/** Generate proxy classes for @main functions.
+ *  A function like
+ *
+ *     @main f(x: S, ys: T*) = ...
```
Comment (hrhino) [inline compiler/src/dotty/tools/dotc/ast/MainProxies.scala]: `@main def`, right?

File: tests/neg/main-functions.scala [hunk 4]
```diff
@@ -0,0 +1,15 @@
+object Test1 {
+  @main def f(x: Foo) = () // error: no implicit argument of type util.FromString[Foo] was found
+}
+
+object Test2 {
+  @main val x = 2  // does nothing, should this be made an error?
```
Comment (hrhino) [inline tests/neg/main-functions.scala]: I'd agree that `@main val` shouldn't be allowed.

File: tests/neg/main-functions.scala [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+}
+
+class Foo {
+  @main def f = () // does nothing, should this be made an error?
```
Comment (hrhino) [inline tests/neg/main-functions.scala]: Probably more trouble than it's worth.

**Comment by soronpo** [comment]
Where can I discuss the deprecation of `App` ?

**Comment by soronpo** [comment]
In what way `@main` is a "special" feature,  and not just a private case for annotation macros (which don't exist yet in dotty)?

File: compiler/src/dotty/tools/dotc/ast/MainProxies.scala (hunk 3 above)
Comment (odersky) [reply compiler/src/dotty/tools/dotc/ast/MainProxies.scala]: right.

**Comment by odersky** [comment]
@soronpo If we had annotation macros `@main` would indeed be expressible with them. But we don't have them, and will probably not get them for 3.0 (maybe later, there are no plans at the moment either way). At the same time we have to do something about the `App` situation. Current `App` requires `DelayedInit` which we don't have either. So we should already promote something else before we go into feature freeze. If `@main` methods can in the future be expressed with macro annotations so much the better!

File: docs/docs/reference/changed-features/main-functions.md [hunk 6]
```diff
@@ -0,0 +1,82 @@
+---
+layout: doc-page
+title: "Main Methods"
+---
+
+Scala 3 offers a new way to define programs that can be invoked from the command line:
+A `@main` annotation on a method turns this method into an executable program.
+Example:
+```scala
+    @main def happyBirthday(age: Int, name: String, others: String*) = {
```
Comment (soronpo) [inline docs/docs/reference/changed-features/main-functions.md]: Should we encourage the use of `: Unit` here or not?

**Comment by soronpo** [comment]
What is stopping us from annotating an `object` instead of/in addition to a `def`?
Currently `App` is treated differently in some `DelayedInit` fashion which I understand we wish to get rid of.
```scala
object MyApp extends App {
  //do something
}
```
I think we should also allow `@main` to work on an object with a `def main` so `App` can remain the same, but without extending `DelayedInit`:
```scala
@main object MyApp extends App {
  //do something
}
```
In my code base I have: 
```scala
 trait DSLApp extends App { 
  //DSL Library stuff here
}

object DSLUserApp extends DSLApp {
  //User stuff here
}
```

I would very much like the ability to extend `App` (or something else), and it does not mean we need to carry the burden of `DelayedInit`.

**Comment by jducoeur** [comment]
Along the lines of @soronpo's point: the FP community makes pretty heavy use of `IOApp`, IIRC.  So this notion of wanting a specialized main entry point is pretty widespread -- we'd be well-advised to provide *some* means of supporting that.

**Comment by odersky** [comment]
`IOApp` is unaffected by this change, you can still use it as is.

The general problem with annotating an object for main is that the main program is the run in the object's initializer code. The JIT tends to be less aggressive in optimizing such code. This is mainly a problem for micro-benchmarks, but that matters, since you get your reputation ruined quickly if people run the same trivial program in Java and Scala and the Scala version is x times slower. `DelayedInit` gets around that by placing the same code in a method, but that seriously contorts semantics, so we want to get rid of it. 

Note that that's not a problem for `IOApp`, because that one requires you to place your code in the `run` method.

**Comment by odersky** [comment]
I forgot the more important reason why we do not want to put initalization code in object initializers: It's prone to deadlocks if your program is multi-threaded.

File: library/src/scala/util/CommandLineParser.scala [hunk 7]
```diff
@@ -0,0 +1,44 @@
+package scala.util
+
+/** A utility object to support command line parsing for @main methods */
+object CommandLineParser {
```
Comment (julienrf) [inline library/src/scala/util/CommandLineParser.scala]: I understand the goal of simplifying the way to implement application entry points. In practice, most of the users just write a `def main(args: Array[String]): Unit` method, and manually parse the application arguments. For these cases, this PR definitely improves the developer experience.

That being said, by introducing a way to parse the program arguments in the compiler, we will also establish a “standard” way of doing that. And, I’m a little bit worried about the introduction of the `CommandLineParser` and `FromString` abstractions. For now, they are simple but I think we can already anticipate the following requests soon: can we also support optional parameters? Can we support named parameters? How could we customize the error message in case of invalid parameters? Can we print all the error messages instead of stopping at the first one? Should we use `Either` instead of relying on exceptions to model parsing failures? All these requests are legitimate, in my opinion, but if we try to fulfill them the size of the `CommandLineParser` and `FromString` modules will be 10x bigger. Then, the question will be “should we keep that thing in the compiler, or should we move it to the library land (like scopt, or optparse-applicative)?”. We can make the decision of keeping `CommandLineParser` simple and not honor these requests, “by design”, but then I’m worried about introducing a “standard” way of parsing command-line arguments which does not scale to advanced use cases.

Unfortunately, I don’t have a simple idea of how to decouple `@main` expansion from `CommandLineParser` (so that users could plug in the parsing + reporting logic they want). I still think the most flexible way to go is to require users to forward a plain old `main` method to whatever they want:

~~~ scala
def happyBirthday(age: Int, name: String, others: String*): Unit = …

def main(args: Array[String]): Unit = {
  val happyBirthdayParser = … // some code that relies on a library for arguments parsing
  happyBirthdayParser.run(args)(happyBirthday)
}
~~~

But I agree that this approach is *not* beginner-friendly…

**Comment by liufengyun** [comment]
I find the following is more natural, no semantic magic with names: 

The code

```Scala
    @main class HappyBirthday(age: Int, name: String, others: String*) {
        // code
    }
```

expands to

```Scala
    class HappyBirthday(age: Int, name: String, others: String*) {
	// code
    }

    object HappyBirthday {
      import scala.util.{CommndLineParser => CLP}
      def main(args: Array[String]): Unit = {
	try
	  new HappyBirthday(
	      CLP.parseArgument[Int](args, 0),
	      CLP.parseArgument[String](args, 1),
	      CLP.parseRemainingArguments[String](args, 2))
	catch {
	  case error: CLP.ParseError => CLP.showError(error)
	}
      }
    }
```

**Comment by liufengyun** [comment]
I don’t see how benchmarking is an issue, as usage of  JMH will requires the code to be in a method.

**Comment by soronpo** [comment]
@odersky is supporting both forms (`@main object ...`, `@main def ...`) out of the question? The latter is used in most cases and the first to ease migration and allow the example I've given above.
Or alternatively if we can do `@main class MyApp extends Array[String] => Unit {}`, that is fine too I believe.

**Comment by odersky** [comment]
> @odersky is supporting both forms (@main object ..., @main def ...) out of the question? 

For me, yes. We don't need more than one convenient way to do main programs. `@main def` has the advantage that it's what ammonite does, so we close the gap between scripting and normal programs.

File: library/src/scala/util/CommandLineParser.scala (hunk 7 above)
Comment (odersky) [reply library/src/scala/util/CommandLineParser.scala]: I believe we can stay firm and keep `CommandLineParser` simple. If one wants something else, one can perfectly well define a traditional main method, or define the entry point like this:
```scala
@main def f(args: String*) = ...
```

File: compiler/src/dotty/tools/dotc/ast/MainProxies.scala [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+
+    var result: List[TypeDef] = Nil
+    if (!mainFun.owner.isStaticOwner)
+      ctx.error(s"@main method is not statically accessible", mainFun.sourcePos)
+    else {
+      var call = ref(mainFun.termRef)
+      mainFun.info match {
+        case _: ExprType =>
+        case mt: MethodType =>
+          if (!mt.isImplicitMethod) call = addArgs(call, mt.paramInfos, mt.resultType, 0)
```
Comment (anatoliykmetyuk) [inline compiler/src/dotty/tools/dotc/ast/MainProxies.scala]:
```scala
object Main {
  delegate for Int = 10
  @main def g given (x: Int): Unit = {
    println(s"Hello with $x")
  }
}
```

Fails with:

```scala
-- Error: ../ecosystem/mainmethods/src/main/scala/Main.scala:11:2 --------------
11 |  @main def g given (x: Int): Unit = {
   |  ^
   |no implicit argument of type Int was found for parameter x of method g in object Main
```

Whereas:

```scala
delegate for Int = 10
object Main {
  @main def g given (x: Int): Unit = {
    println(s"Hello with $x")
  }
}
```

Works fine. I suspect this has to do with the fact that `g` is a static method in a class `g` and somehow the inner `delegate` becomes invisible to it. Even in that scenario though, it should be visible IMO.

File: docs/docs/reference/changed-features/main-functions.md [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+            CLP.parseArgument[Int](args, 0),
+            CLP.parseArgument[String](args, 1),
+            CLP.parseRemainingArguments[String](args, 2))
+      catch {
+        case error: CLP.ParseError => CLP.showError(error)
+      }
+  }
+```
+**Note**: The `<static>` modifier above expresses that the `main` method is generated
+as a static method of class `happyBirthDay`. It is not available for user programs in Scala. Regular "static" members are generated in Scala using objects instead.
```
[the remark this answers is not in the document]
Comment (anatoliykmetyuk) [reply docs/docs/reference/changed-features/main-functions.md]: > It is not available for user programs in Scala

Also, it doesn't seem that this limitation is true:

```scala
object Main {
  def main(args: Array[String]): Unit = { f(); println("foo") }

  @main def f(): Unit = {
    println("Hello world!")
    println(msg)
  }

  @main def g(): Unit = f()

  def msg = "I was compiled by dotty :)"

}
```

Compiles and runs fine.

File: compiler/src/dotty/tools/dotc/ast/MainProxies.scala (hunk 8 above)
Comment (odersky) [reply compiler/src/dotty/tools/dotc/ast/MainProxies.scala]: The behavior is actually intended, and follows from the way expansion of @main is specified.
The method `g` is called from a synthetic class `g` which is conceptually in the same compilation unit as `g` but is at the toplevel. So it an see a global import but it cannot see a local given instance.

File: library/src/scala/util/FromString.scala [hunk 10]
```diff
@@ -0,0 +1,47 @@
+package scala.util
+
+trait FromString[T] {
```
[the remark this answers is not in the document]
Comment (odersky) [reply library/src/scala/util/FromString.scala]: `Option` is clearly not an adequate replacement, since it does not allow to pass an error string _why_ the argument does not conform. 

Also, in the end my user-level method takes, say an `Int` and `Boolean`. If the arguments do not conform to that I _must_ throw an exception. There's no other way to do it. Unless you make @main programs take only `Option` or `Either` arguments, but that's completely wrong! The whole point of @main functions is that the person writing the @main can decide what degree of error checking should be done, and how. If you force @main functions to handle errors by analyzing `Option` or `Either` arguments you made the choice for them.

In summary: exceptions are actually pretty nice! And we will fix the part that they don't show up in the types...

**Comment by smarter** [comment]
> I believe sbt may not be prepared to the fact that main methods may be static and/or come from classes and not objects (i.e. classes that end with $).

We control main class detection, not sbt: https://github.com/lampepfl/dotty/blob/232b3fa9e17a771e55088299d5e461ee25aed9e5/compiler/src/dotty/tools/dotc/sbt/ExtractAPI.scala#L233-L235

**Comment by odersky** [comment]
Thanks for the hint @smarter! I tried a fix. @anatoliykmetyuk Can you test whether it works now with sbt?

File: compiler/src/dotty/tools/dotc/sbt/ExtractAPI.scala [hunk 11]
```diff
@@ -230,7 +230,7 @@ private class ExtractAPICollector(implicit val ctx: Context) extends ThunkHolder
 
     allNonLocalClassesInSrc += cl
 
-    if (sym.isStatic && defType == DefinitionType.Module && ctx.platform.hasMainMethod(sym)) {
+    if (sym.isStatic && ctx.platform.hasMainMethod(sym)) {
```
Comment (allanrenucci) [inline compiler/src/dotty/tools/dotc/sbt/ExtractAPI.scala]: I think you need to drop `sym.isStatic` to allow classes to have `main` method.

**Comment by smarter** [comment]
We should add a scripted sbt test to verify the main class detection work: https://github.com/lampepfl/dotty/tree/master/sbt-dotty/sbt-test/sbt-dotty

**Comment by anatoliykmetyuk** [comment]
As of `8b2737d`, SBT detects the `@main` methods

File: docs/docs/reference/changed-features/main-functions.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (hrhino) [reply docs/docs/reference/changed-features/main-functions.md]: Doesn't that mean that the _synthetic_ method `main` can't be called, rather than that the `@main` method written in source can't be called?

File: compiler/src/dotty/tools/dotc/ast/MainProxies.scala (hunk 8 above)
Comment (hrhino) [reply compiler/src/dotty/tools/dotc/ast/MainProxies.scala]: That makes sense, but it seems strange that the compile-time behavior is changed by the annotation.

Why do we allow implicit parameters on `@main` methods? I suppose "it resolves whatever it needs at the definition site" is the only semantics that makes sense, but I don't know another place in Scala where the definition-site implicit scope affects anything. Or can the implicit parameters participate in command-line parsing?

Comment (anatoliykmetyuk) [reply compiler/src/dotty/tools/dotc/ast/MainProxies.scala]: The problem is:

```scala
object Main {
  delegate for Int = 10
  @main def g given (x: Int): Unit = {
    println(s"Hello with $x")
  }
}
```

Desugars to:

```scala
object Main {
  delegate for Int = 10
  def g given (x: Int): Unit = {
    println(s"Hello with $x")
  }
}

class g {
  <static> def main(args: Array[String]): Unit = g(...)
}
```

The implicits defined at `Main` are not resolved from `g` because `class g` is ejected from `Main`. We cannot have it nested because `main` method must be in a top-level class.

However:

```scala
object Main {
  delegate for Int = 10
  def g given (x: Int): Unit = {
    println(s"Hello with $x")
  }
  def g$main$implicitless: Unit = g
}

class g {
  <static> def main(args: Array[String]): Unit = g$main$implicitless
}
```

@odersky do you think something like that will work to mitigate the unintuitive behaviour of the implicits with nested `@main`s?

Comment (odersky) [reply compiler/src/dotty/tools/dotc/ast/MainProxies.scala]: We need to have a boringly simple and obvious mapping from @main function to the generated main class I think that the implemented behavior is not unintuitive at all. The generated main class comes at the end of the compilation unit that contains the @main function. That's all one needs to explain. From that it follows that global imports are visible but local imports are not.

Comment (anatoliykmetyuk) [reply compiler/src/dotty/tools/dotc/ast/MainProxies.scala]: In the current implementation, the local imports are not a problem since they will be visible as per the users' intuition:

```scala
object foo {
  val x: Int = 11
}

object Main {
  import foo._
  delegate for Int = 10
  @main def g: Unit = { //given (x: Int): Unit = {
    println(s"Hello with ${the[Int]} and $x")
  }
}
```

Because the imports are used from the _body_ of the @main function, and the function itself does not change its location. The only problem is resolving the implicits – this is because they present the only resolution which happens on the _call_ of the function.  And the call happens in some other location. As far as I know that the only resolution that happens not in the body of the function.

> We need to have a boringly simple and obvious mapping from @main function to the generated main class.

In the current case, we need to teach the users that the behaviour will be unintuitive with respect to implicits (while their scoping intuition will hold for everything else). In the solution I mentioned we will need to tech that a bridge is generated for the main methods to support proper implicits scoping.

In both cases we need to teach them a subtlety about the language. However, if we do generate bridges for implicits, we ensure that the users' intuition holds.

Comment (anatoliykmetyuk) [reply compiler/src/dotty/tools/dotc/ast/MainProxies.scala]: I think the generated main function should have only the following body:

```scala
<static> def main(args: Array[String]): Unit = path.to.`main's`.parent.object.main$no$implcits(args)

// In that object
def main$no$implcits(args: Array[String]) = {  // copied from `MainProxies.scala`
          try
            f(
              parseArgument[S](args, 0),
              parseRemainingArguments[T](args, 1): _*
            )
          catch case err: ParseError => showError(err)
}
```

This will ensure that also the nested implicit `FromString` objects are resolved properly.

**Comment by odersky** [comment]
@anatoliykmetyuk I think everything we agreed on is done. Merge?
